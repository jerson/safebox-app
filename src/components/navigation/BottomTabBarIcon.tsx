import React from 'react';
import {
  StyleSheet,
  TouchableWithoutFeedback,
  View,
  ViewStyle,
  StyleProp,
  Animated
} from 'react-native';
import Colors from '../../modules/constants/Colors';

const styles = StyleSheet.create({
  background: {
    position: 'absolute',
    left: 10,
    right: 10,
    top: 5,
    bottom: -10,

    borderTopLeftRadius: 100,
    borderTopRightRadius: 100,
    backgroundColor: Colors.primaryLight,
    zIndex: 0
  } as ViewStyle,
  content: {
    flex: 1,
    position: 'relative'
  } as ViewStyle
});

export interface BottomTabBarIconProps {
  focused: boolean;
  onPress: () => void;
  onLongPress: () => void;
  accessibilityLabel: string;
  style: StyleProp<ViewStyle>;
}

export interface BottomTabBarIconState {
  active: Animated.Value;
}

export default class BottomTabBarIcon extends React.Component<
  BottomTabBarIconProps,
  BottomTabBarIconState
> {
  state: BottomTabBarIconState = {
    active: new Animated.Value(0)
  };

  animation!: Animated.CompositeAnimation;

  componentDidMount() {
    const { focused } = this.props;

    this.setState({ active: new Animated.Value(focused ? 1 : 0) });
  }

  componentDidUpdate(prevProps: BottomTabBarIconProps) {
    const { focused } = this.props;
    if (prevProps.focused !== focused) {
      if (focused) {
        this.focus();
      } else {
        this.blur();
      }
    }
  }

  focus() {
    this.animation && this.animation.stop();
    this.animation = Animated.spring(this.state.active, {
      toValue: 1,
      friction: 4
    });

    this.animation.start();
  }

  blur() {
    this.animation && this.animation.stop();
    this.animation = Animated.spring(this.state.active, {
      toValue: 0
    });
    this.animation.start();
  }

  render() {
    const {
      accessibilityLabel,
      onPress,
      onLongPress,
      focused,
      children,
      ...props
    } = this.props;
    const { active } = this.state;

    const opacity = active.interpolate({
      inputRange: [0, 0.2, 1],
      outputRange: [0, 1, 1]
    });
    const scale = active.interpolate({
      inputRange: [0, 1],
      outputRange: [0.1, 1]
    });
    const top = active.interpolate({
      inputRange: [0, 1],
      outputRange: [55, 2]
    });

    const transform = {
      opacity,
      top,
      transform: [
        {
          scale
        }
      ]
    };

    return (
      <TouchableWithoutFeedback
        accessibilityLabel={accessibilityLabel}
        onPress={onPress}
        onLongPress={onLongPress}
      >
        <View style={styles.content}>
          <Animated.View style={[styles.background, transform]} />

          <View {...props}>{children}</View>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}
