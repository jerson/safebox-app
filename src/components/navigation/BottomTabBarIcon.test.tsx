import React from 'react';
import * as renderer from 'react-test-renderer';

import BottomTabBarIcon, {BottomTabBarIconProps} from './BottomTabBarIcon';

const props: BottomTabBarIconProps = {
  focused: true,
  accessibilityLabel: 'test',
  onLongPress: jest.fn(),
  onPress: jest.fn(),
  style: {},
};
test('renders without crashing', () => {
  const snapshot = renderer.create(<BottomTabBarIcon {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<BottomTabBarIcon {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
