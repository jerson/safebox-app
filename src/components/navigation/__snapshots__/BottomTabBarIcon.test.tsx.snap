// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<_class
  accessibilityLabel="test"
  accessible={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "position": "relative",
    }
  }
>
  <_class
    style={
      Object {
        "backgroundColor": "#9066f5",
        "borderTopLeftRadius": 100,
        "borderTopRightRadius": 100,
        "bottom": -60,
        "left": 10,
        "opacity": 1,
        "position": "absolute",
        "right": 10,
        "top": 2,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "zIndex": 0,
      }
    }
  />
  <_class
    focused={true}
    style={Object {}}
  />
</_class>
`;
